### DISCUSSION TITLE
How to automatically cut songs that have a long silence and a "bonus track" at the end

### DISCUSSION THREAD

**Discussion by Moonbase59** [AzuraCast/AzuraCast#6334] [Show and tell]
Some songs have a long silence, sometimes several minutes, at their end, followed by what they call a "bonus track". In Radio, you’d want to just play the song and skip the silence and the "bonus"—we don’t want our stations to be silent for several minutes, after all!

Fortunately, Liquidsoap offers an easy way to do just that: Skip a song if it goes below a certain power level, and continue with the next one.

In your Liquidsoap configuration, in the input box just _above_ the

```
def live_aware_crossfade(old, new) =
```

enter the following:

```
# If a song has <song> <silence> <bonus track>,
# blank.skip() will skip to next song and NOT play <bonus track>.
# Defaults: max_blank=20.0 (seconds), threshold=-40.0 (dB)
# I found 5s silence and a threshold of -60 dB quite usable.
radio = blank.skip(radio, max_blank=5.0, threshold=-60.0)
```

As an example, take the song _Walk With Me_ by _The Tea Party_, from their 1995 album _The Edges of Twilight_: It has a total length of 14:14, starts rather low-volume and ends around 7:25. This is followed by about 33s of silence, followed by 2:10 of song, then another silence of 3:45, then some sound at about 14:00!

![Auswahl_220](https://github.com/AzuraCast/AzuraCast/assets/3706922/e581aa47-f9ba-40e9-a6b8-51b2122fd861)

Using above setting, this song will be ended at 7:25 and your station smoothly continue with the next song. Ain’t Liquidsoap great?

**Note:** If using sources that have correct replaygain tags, you should apply replaygain _before_ applying `blank.skip()`, for instance like so:

```
# Be sure to have ReplayGain applied before crossing.
radio = amplify(1.,override="replaygain_track_gain",radio)

# If a song has <song> <silence> <bonus track>,
# blank.skip() will skip to next song and NOT play <bonus track>.
# Defaults: max_blank=20.0 (seconds), threshold=-40.0 (dB)
radio = blank.skip(radio, max_blank=5.0, threshold=-60.0)
```

**Note:** If using a custom crossfade, both of above should come _before_ your custom crossfade.

Happy playout!

**Comment by blindndangerous** [comment]
You are awesome, thank you for all of these snippets. It's really helpful.

**Comment by Moonbase59** [comment]
Thank you! The _real_ awesomeness lies of course by the people _making_ things like AzuraCast, Liquidsoap, ffmpeg and Icecast, and I thank them with all my heart for making all that available to us as FOSS!

**Comment by BornKillaz** [comment]
Amazing! Thank you so much for sharing this.

**Comment by Moonbase59** [comment]
Nowadays (May 2024) you can also use my `autocue.cue_file` integration for this. Just use this setting:

```
settings.autocue.cue_file.blankskip := true
```

We can verify the result by using `cue_file` on the command line for above song:

```
$ cue_file -b "Tea Party, The - Walk With Me.flac" | jq -S
{
  "duration": 854.2,
  "liq_amplify": "-7.10 dB",
  "liq_amplify_adjustment": "0.00 dB",
  "liq_blank_skipped": true,
  "liq_blankskip": true,
  "liq_cross_start_next": 437,
  "liq_cue_duration": 443.90000000000003,
  "liq_cue_in": 1.7,
  "liq_cue_out": 445.6,
  "liq_longtail": false,
  "liq_loudness": "-10.90 LUFS",
  "liq_loudness_range": "11.10 LU",
  "liq_reference_loudness": "-18.00 LUFS",
  "liq_true_peak": "0.22 dBFS"
}
```

As you can see, `cue_file` determined a total song duration of 854.2s (14:14) and set the cue-out point at 445.6s (7:26), starting the next song at 437s (7:17) for a nice transition!

![Screenshot of waveform, showing _Start next song_ and _Cue-out_ points; overlay area marked.](https://github.com/AzuraCast/AzuraCast/assets/3706922/2e6fdf3e-81d2-4890-b29d-c6c19ef087e8)

**Comment by BornKillaz** [comment]
Hi!
What's the advantage of using this method instead of the other one? Do I just need to paste that line on liquidsoap configuration and will work out of the box?

settings.autocue.cue_file.blankskip := true

Also, is it possible to set this feature to apply only to audio files shorter than 45 minutes, and disable it for files longer than 45 minutes? That would be awesome, as we have some pre-recorded shows and I don't want a 2 hours show to be cut off by this, just in case. 

Thank you.

**Comment by Moonbase59** [comment]
Well the autocues are a work in progress, and mainly intended for making nicer transitions between songs. Since many of us have songs with "hidden tracks" and don ’t want to destroy their originals, I built that function in.

It all depends on your use case. You can disable my `autocue.cue_file` completely, or only the blank skipping function on a per-file (tags/annotation) or per-playlist (annotation) basis, but that’s not (yet?) integrated into AzuraCast.

- `liq_cue_file=false` will disable my autocue (for a file/playlist/single)
- `liq_blankskip=false` will disable blank skipping for a file/playlist/single
- AzuraCast jingle playlists that have _Hide Metadata from Listeners ("Jingle Mode")_ set to ON will automatically switch off blank skipping for jingles.

You’d have to install my autocue integration to use it, including the `cue_file` executable. So if you don’t want/need the other functions, it _could_ be a little overkill for just skipping blanks.

See also: https://github.com/AzuraCast/AzuraCast/discussions/6252#discussioncomment-9344857
